domain/controllerconfig/state/state.go

```diff
@@ -56,44 +68,96 @@ func (st *State) ControllerConfig(ctx context.Context) (map[string]string, error
 // after bootstrapping.
 // ValidateModification is a function that will be called with the current
 // config, and should return an error if the modification is not allowed.
-func (st *State) UpdateControllerConfig(ctx context.Context, updateAttrs map[string]string, removeAttrs []string, validateModification func(map[string]string) error) error {
+func (st *State) UpdateControllerConfig(
+	ctx context.Context,
+	updateAttrs map[string]string, removeAttrs []string,
+	validateModification func(map[string]string) error,
+) error {
 	db, err := st.DB()
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	selectQuery := "SELECT key, value FROM controller_config"
-	deleteQuery := "DELETE FROM controller_config WHERE key = ?"
-	updateQuery := `
-INSERT INTO controller_config (key, value)
-VALUES (?, ?)
-  ON CONFLICT(key) DO UPDATE SET value=?`
+	selectStmt, err := st.Prepare("SELECT &dbKeyValue.* FROM v_controller_config", dbKeyValue{})
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
-		// Check keys and values are valid between current and new config.
-		rows, err := tx.QueryContext(ctx, selectQuery)
+	deleteStmt, err := st.Prepare("DELETE FROM controller_config WHERE key IN ($S[:])", sqlair.S{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	removeKeys := sqlair.S(transform.Slice(removeAttrs, func(s string) any { return any(s) }))
```

Comment:
I've used a []string type alias and avoided the transform - to me it's cleaner that way. YMMV.